ldap/servers/slapd/dyncerts.c

```diff
@@ -798,6 +797,786 @@ dyncerts_cert2entry(CERTCertificate *cert)
+            return LDAP_OBJECT_CLASS_VIOLATION;
+        }
+    }
+    /* Check that nsDynamicCertificateDER attribute is present */
+    if ((needcert || slapi_entry_attr_get_ref(e, DYCATTR_PKEYDER)) &&
+        !slapi_entry_attr_get_ref(e, DYCATTR_CERTDER)) {
+        PR_snprintf(errmsg, SLAPI_DSE_RETURNTEXT_SIZE,
+                    "Attribute %s should be present", DYCATTR_CERTDER);
+        return LDAP_UNWILLING_TO_PERFORM;
+    }
+    return rc;
+}
+
+/* Store the entry in NSS db */
+int
+dyncerts_import_entry(Slapi_Entry *e, char *errmsg, bool verifyOnly)
+{
+    int rc = LDAP_SUCCESS;
+    CertCtx_t ctx = {0};
+
+    dyncert_nickname_from_full_nickname(&ctx.n, slapi_entry_attr_get_ref(e, DYCATTR_NICKNAME));
+    ctx.dercert = slapi_entry_attr_get_secitem(e, DYCATTR_CERTDER);
+    ctx.derpkey = slapi_entry_attr_get_secitem(e, DYCATTR_PKEYDER);
+    ctx.trust = slapi_entry_attr_get_charptr(e, DYCATTR_TRUST);
+    ctx.force = (slapi_entry_attr_get_bool(e, DYCATTR_FORCE) == PR_TRUE);
+    ctx.errmsg = errmsg;
+    ctx.sdn = (Slapi_DN*) slapi_entry_get_sdn_const(e);
+    rc = dyncerts_import_cert_and_key(&ctx, verifyOnly);
+    dyncerts_cert_ctx_done(&ctx);
+    return rc;
+}
+
+/* Backend callback (add operation) */
+int
+dyncerts_add(Slapi_PBlock *pb)
+{
+    char returntext[SLAPI_DSE_RETURNTEXT_SIZE] = "";
+    struct dyncerts *pdcerts = NULL;
+    int error = LDAP_SUCCESS;
+    int rc = LDAP_SUCCESS;
+    Slapi_Entry *e = NULL;
+    const char *dn = "???";
+    Slapi_DN *sdn = NULL;
+    Nickname_t n = {0};
+    DCSS *ss = NULL;
+
+    /*
+     * Get the database, the dn and the entry to add
+     */
+    if (slapi_pblock_get(pb, SLAPI_PLUGIN_PRIVATE, &pdcerts) < 0 ||
+        slapi_pblock_get(pb, SLAPI_ADD_TARGET_SDN, &sdn) < 0 ||
+        slapi_pblock_get(pb, SLAPI_ADD_ENTRY, &e) < 0 || (NULL == pdcerts)) {
+        rc = LDAP_OPERATIONS_ERROR;
+        goto done;
+    }
+    dn = slapi_sdn_get_dn(slapi_entry_get_sdn_const(e));
+
+    /* Check entry attributes */
+    (void) dyncert_nickname_from_dn(&n, sdn);
+    rc = dyncerts_check_entry(pb, e, &n, returntext, true);
+    if (rc) {
+        e = NULL; /* caller will free upon error */
+        goto done;
+    }
+    /* Check that entry does not exist */
+    if (dyncerts_find_entry(sdn, LDAP_SCOPE_BASE, NULL, &ss)) {
+        e = NULL; /* caller will free upon error */
+        rc = LDAP_ALREADY_EXISTS;
+        goto done;
+    }
+    rc = dyncerts_import_entry(e, returntext, false);
+done:
+    /* make sure OPRETURN and RESULT_CODE are set */
+    slapi_pblock_get(pb, SLAPI_PLUGIN_OPRETURN, &error);
+    if (rc) {
+        if (!error) {
+            slapi_pblock_set(pb, SLAPI_PLUGIN_OPRETURN, &rc);
+        }
+        slapi_log_err(SLAPI_LOG_DYC, "dyncerts_add",
+                      "Add operation of entry %s failed: rc=%d : %s\n",
+                      dn, rc, returntext);
+    }
+    ss_destroy(&ss);
+    free_config_info();
```

Comment:
If we raise this point before calling dyncerts_import_entry (goto done), the mutex was not acquire yet.